### PR TITLE
Neo datetime: dynamic field loading based on pattern

### PR DESCRIPTION
**Pull request by sffc**
Part of #3347, #3865

### CODE REVIEW

File: components/datetime/src/format/datetime.rs [hunk 1]
```diff
@@ -573,6 +573,112 @@ impl RequiredData {
...
+    ) -> Result<bool, Field> {
+        let fields = pattern.items.iter().filter_map(|p| match p {
+            PatternItem::Field(field) => Some(field),
+            _ => None,
+        });
+
+        enum NeedsWhat<'a> {
+            Nothing,
+            Symbols(&'a mut Option<Field>),
+            FixedDecimalFormatter,
```
Comment (Manishearth) [inline components/datetime/src/format/datetime.rs]: praise: nice, clean, and extensible

File: components/datetime/src/format/datetime.rs [hunk 2]
```diff
@@ -573,6 +573,112 @@ impl RequiredData {
     }
 }
 
+/// What data is required to format a given pattern.
+#[derive(Default)]
+pub(crate) struct RequiredDataNeo {
+    // C::YearSymbolsV1Marker is required.
+    pub year_symbols: Option<Field>,
```
Comment (Manishearth) [inline components/datetime/src/format/datetime.rs]: issue: we should document the invariant that the field must be a year field (etc)

File: components/datetime/src/format/neo.rs [hunk 3]
```diff
@@ -648,6 +655,47 @@ impl<C: CldrCalendar> TypedDateTimePatternInterpolator<C> {
...
+            + ?Sized,
+    {
+        let mut required_data = RequiredDataNeo::default();
+        // TODO: Consider support for time zones here
+        required_data.add_requirements_from_pattern(pattern, false);
+        if let Some(field) = required_data.year_symbols {
+            self.load_year_names(provider, field.length)?;
+        }
+        if let Some(field) = required_data.month_symbols {
+            let FieldSymbol::Month(month) = field.symbol else {
```
Comment (Manishearth) [inline components/datetime/src/format/neo.rs]: nit: i don't think we should panic in release mode here.

and i'm not entirely happy about the invariant existing in the first place, but can't be helped

**Review by Manishearth — APPROVED** [review]

File: components/datetime/src/format/datetime.rs (hunk 2 above)
Comment (sffc) [reply components/datetime/src/format/datetime.rs]: Already fixed in f53809a

File: components/datetime/src/format/neo.rs (hunk 3 above)
Comment (sffc) [reply components/datetime/src/format/neo.rs]: Already fixed in f53809a

File: components/datetime/src/format/datetime.rs (hunk 1 above)
Comment (sffc) [reply components/datetime/src/format/datetime.rs]: Thanks, but I inlined this function directly into DateTimePatternInterpolator in f53809a which is not quite as elegant but it has fewer invariants between modules which I think is better overall.

**Pull request merged**